### ISSUE TITLE
Significant inefficiency in TagsToShortcode

### ISSUE THREAD

**Issue by sminnee** [silverstripe/silverstripe-assets#316]
TagsToShortcode seems to re-execute the SiteTree.Content rewrite once for every subclass of SiteTree, on the same table.

Below is the `$classes` array in a candidate execution, of the form`[ $class => [ $table => [ ...$fields ] ] ]`. It's generated by TagsToShortcodeHelper::getFieldMap()

One possible fix is for getFieldMap() to have a pruning pass over the result before returning its value, where subsequent appearances of tables are removed.

This task can take hours to run, so it'd be nice to speed it up

```
array(35) {
  ["SilverStripe\CMS\Model\SiteTree"]=>
  array(1) {
    ["SiteTree"]=>
    array(1) {
      [0]=>
      string(7) "Content"
    }
  }
  ["Page"]=>
  array(2) {
    ["SiteTree"]=>
    array(1) {
      [0]=>
      string(7) "Content"
    }
    ["Page"]=>
    array(2) {
      [0]=>
      string(12) "RightContent"
      [1]=>
      string(13) "TitleImageAlt"
    }
  }
  ["SilverStripe\Org\Component\ContentComponent"]=>
  array(1) {
    ["SiteTree"]=>
    array(1) {
      [0]=>
      string(7) "Content"
    }
  }
  ["BluehouseGroup\Pardot\PardotShortCode"]=>
  array(1) {
    ["SiteTree"]=>
    array(1) {
      [0]=>
      string(7) "Content"
    }
  }
  ["SilverStripe\Org\PageType\CommunityPage"]=>
  array(2) {
    ["SiteTree"]=>
    array(1) {
      [0]=>
      string(7) "Content"
    }
    ["Page"]=>
    array(2) {
      [0]=>
      string(12) "RightContent"
      [1]=>
      string(13) "TitleImageAlt"
    }
  }
  ["SilverStripe\Org\PageType\CommunityShowcasePage"]=>
  array(3) {
    ["SiteTree"]=>
    array(1) {
      [0]=>
      string(7) "Content"
    }
    ["Page"]=>
    array(2) {
      [0]=>
      string(12) "RightContent"
      [1]=>
      string(13) "TitleImageAlt"
    }
    ["CommunityShowcasePage"]=>
    array(2) {
      [0]=>
      string(14) "SuccessContent"
      [1]=>
      string(17) "SubmissionContent"
    }
  }
  ["SilverStripe\Org\PageType\ContributeHolder"]=>
  array(2) {
    ["SiteTree"]=>
    array(1) {
      [0]=>
      string(7) "Content"
    }
    ["Page"]=>
    array(2) {
      [0]=>
      string(12) "RightContent"
      [1]=>
      string(13) "TitleImageAlt"
    }
  }
  ["SilverStripe\Org\PageType\ContributePage"]=>
  array(2) {
    ["SiteTree"]=>
    array(1) {
      [0]=>
      string(7) "Content"
    }
    ["Page"]=>
    array(2) {
      [0]=>
      string(12) "RightContent"
      [1]=>
      string(13) "TitleImageAlt"
    }
  }
  ["SilverStripe\Org\PageType\DeveloperNetworkPage"]=>
  array(3) {
    ["SiteTree"]=>
    array(1) {
      [0]=>
      string(7) "Content"
    }
    ["Page"]=>
    array(2) {
      [0]=>
      string(12) "RightContent"
      [1]=>
      string(13) "TitleImageAlt"
    }
    ["DeveloperNetworkPage"]=>
    array(3) {
      [0]=>
      string(15) "JoinRequirement"
      [1]=>
      string(9) "IntroText"
      [2]=>
      string(17) "SubmissionContent"
    }
  }
  ["SilverStripe\Org\PageType\DownloadPage"]=>
  array(3) {
    ["SiteTree"]=>
    array(1) {
      [0]=>
      string(7) "Content"
    }
    ["Page"]=>
    array(2) {
      [0]=>
      string(12) "RightContent"
      [1]=>
      string(13) "TitleImageAlt"
    }
    ["DownloadPage"]=>
    array(1) {
      [0]=>
      string(15) "PlatformContent"
    }
  }
  ["SilverStripe\Org\PageType\GettingStartedPage"]=>
  array(2) {
    ["SiteTree"]=>
    array(1) {
      [0]=>
      string(7) "Content"
    }
    ["Page"]=>
    array(2) {
      [0]=>
      string(12) "RightContent"
      [1]=>
      string(13) "TitleImageAlt"
    }
  }
  ["SilverStripe\Org\PageType\HelpAndSupportItemPage"]=>
  array(2) {
    ["SiteTree"]=>
    array(1) {
      [0]=>
      string(7) "Content"
    }
    ["Page"]=>
    array(2) {
      [0]=>
      string(12) "RightContent"
      [1]=>
      string(13) "TitleImageAlt"
    }
  }
  ["SilverStripe\Org\PageType\HelpSupportPage"]=>
  array(2) {
    ["HelpSupportPage"]=>
    array(7) {
      [0]=>
      string(7) "Content"
      [1]=>
      string(8) "Abstract"
      [2]=>
      string(8) "Content2"
      [3]=>
      string(8) "Content3"
      [4]=>
      string(8) "Content4"
      [5]=>
      string(8) "Content5"
      [6]=>
      string(8) "Content6"
    }
    ["Page"]=>
    array(2) {
      [0]=>
      string(12) "RightContent"
      [1]=>
      string(13) "TitleImageAlt"
    }
  }
  ["SilverStripe\Org\PageType\LessonPage"]=>
  array(3) {
    ["SiteTree"]=>
    array(1) {
      [0]=>
      string(7) "Content"
    }
    ["Page"]=>
    array(2) {
      [0]=>
      string(12) "RightContent"
      [1]=>
      string(13) "TitleImageAlt"
    }
    ["LessonPage"]=>
    array(1) {
      [0]=>
      string(11) "LessonIntro"
    }
  }
  ["SilverStripe\Org\PageType\SecurityIssueHolder"]=>
  array(2) {
    ["SiteTree"]=>
    array(1) {
      [0]=>
      string(7) "Content"
    }
    ["Page"]=>
    array(2) {
      [0]=>
      string(12) "RightContent"
      [1]=>
      string(13) "TitleImageAlt"
    }
  }
  ["SilverStripe\Org\PageType\SecurityIssuePage"]=>
  array(2) {
    ["SiteTree"]=>
    array(1) {
      [0]=>
      string(7) "Content"
    }
    ["Page"]=>
    array(2) {
      [0]=>
      string(12) "RightContent"
      [1]=>
      string(13) "TitleImageAlt"
    }
  }
  ["SilverStripe\Org\PageType\SlackSignupPage"]=>
  array(2) {
    ["SiteTree"]=>
    array(1) {
      [0]=>
      string(7) "Content"
    }
    ["Page"]=>
    array(2) {
      [0]=>
      string(12) "RightContent"
      [1]=>
      string(13) "TitleImageAlt"
    }
  }
  ["SilverStripe\ErrorPage\ErrorPage"]=>
  array(2) {
    ["SiteTree"]=>
    array(1) {
      [0]=>
      string(7) "Content"
    }
    ["Page"]=>
    array(2) {
      [0]=>
      string(12) "RightContent"
      [1]=>
      string(13) "TitleImageAlt"
    }
  }
  ["Firesphere\StripeSlack\Page\StripeSlackPage"]=>
  array(3) {
    ["SiteTree"]=>
    array(1) {
      [0]=>
      string(7) "Content"
    }
    ["Page"]=>
    array(2) {
      [0]=>
      string(12) "RightContent"
      [1]=>
      string(13) "TitleImageAlt"
    }
    ["StripeSlackPage"]=>
    array(2) {
      [0]=>
      string(7) "Success"
      [1]=>
      string(5) "Error"
    }
  }
  ["SilverStripe\Blog\Model\Blog"]=>
  array(3) {
    ["SiteTree"]=>
    array(1) {
      [0]=>
      string(7) "Content"
    }
    ["Page"]=>
    array(2) {
      [0]=>
      string(12) "RightContent"
      [1]=>
      string(13) "TitleImageAlt"
    }
    ["Blog"]=>
    array(2) {
      [0]=>
      string(13) "SignupContent"
      [1]=>
      string(18) "SignupModalContent"
    }
  }
  ["SilverStripe\Blog\Model\BlogPost"]=>
  array(3) {
    ["SiteTree"]=>
    array(1) {
      [0]=>
      string(7) "Content"
    }
    ["Page"]=>
    array(2) {
      [0]=>
      string(12) "RightContent"
      [1]=>
      string(13) "TitleImageAlt"
    }
    ["BlogPost"]=>
    array(1) {
      [0]=>
      string(7) "Summary"
    }
  }
  ["SilverStripe\CMS\Model\RedirectorPage"]=>
  array(2) {
    ["SiteTree"]=>
    array(1) {
      [0]=>
      string(7) "Content"
    }
    ["Page"]=>
    array(2) {
      [0]=>
      string(12) "RightContent"
      [1]=>
      string(13) "TitleImageAlt"
    }
  }
  ["SilverStripe\CMS\Model\VirtualPage"]=>
  array(2) {
    ["SiteTree"]=>
    array(1) {
      [0]=>
      string(7) "Content"
    }
    ["Page"]=>
    array(2) {
      [0]=>
      string(12) "RightContent"
      [1]=>
      string(13) "TitleImageAlt"
    }
  }
  ["SilverStripe\Org\PageType\HelpAndSupportItemRedirectorPage"]=>
  array(2) {
    ["SiteTree"]=>
    array(1) {
      [0]=>
      string(7) "Content"
    }
    ["Page"]=>
    array(2) {
      [0]=>
      string(12) "RightContent"
      [1]=>
      string(13) "TitleImageAlt"
    }
  }
  ["SilverStripe\Org\Component\CallToActionsComponent"]=>
  array(1) {
    ["SiteTree"]=>
    array(1) {
      [0]=>
      string(7) "Content"
    }
  }
  ["SilverStripe\Org\Component\DownloadComponent"]=>
  array(1) {
    ["SiteTree"]=>
    array(1) {
      [0]=>
      string(7) "Content"
    }
  }
  ["SilverStripe\Org\Component\LatestPostsComponent"]=>
  array(1) {
    ["SiteTree"]=>
    array(1) {
      [0]=>
      string(7) "Content"
    }
  }
  ["SilverStripe\Org\Component\MeetupComponent"]=>
  array(1) {
    ["SiteTree"]=>
    array(1) {
      [0]=>
      string(7) "Content"
    }
  }
  ["SilverStripe\Org\Component\SocialMediaComponent"]=>
  array(1) {
    ["SiteTree"]=>
    array(1) {
      [0]=>
      string(7) "Content"
    }
  }
  ["SilverStripe\Org\Component\StatsComponent"]=>
  array(1) {
    ["SiteTree"]=>
    array(1) {
      [0]=>
      string(7) "Content"
    }
  }
  ["SilverStripe\Org\Component\TableComponent"]=>
  array(1) {
    ["SiteTree"]=>
    array(1) {
      [0]=>
      string(7) "Content"
    }
  }
  ["SilverStripe\Org\Component\AccordionComponent"]=>
  array(1) {
    ["SiteTree"]=>
    array(1) {
      [0]=>
      string(7) "Content"
    }
  }
  ["SilverStripe\Org\Component\AddonsComponent"]=>
  array(1) {
    ["SiteTree"]=>
    array(1) {
      [0]=>
      string(7) "Content"
    }
  }
  ["SilverStripe\Org\Component\ReviewComponent"]=>
  array(1) {
    ["SiteTree"]=>
    array(1) {
      [0]=>
      string(7) "Content"
    }
  }
  ["SilverStripe\Org\Component\TabsComponent"]=>
  array(1) {
    ["SiteTree"]=>
    array(1) {
      [0]=>
      string(7) "Content"
    }
  }
}
```

**Comment by maxime-rainville** [comment]
@sminnee Want to try this one https://github.com/silverstripe/silverstripe-assets/pull/319?

**Comment by robbieaverill** [comment]
Fixed in assets 1.4.x - thanks @maxime-rainville!
